packages/core/src/indexing/sqlite/chronological-order.ts

```diff
@@ -0,0 +1,184 @@
+) {
+  if (!input) return undefined
+  return { created_at: input.created_at, last_anchored_at: input.last_anchored_at }
+}
+
+const CHRONOLOGICAL_ORDER = [
+  { column: 'last_anchored_at', order: 'DESC', nulls: 'last' }, // (`last_anchored_at` is null) DESC
+  { column: 'last_anchored_at', order: 'DESC' },
+  { column: 'created_at', order: 'DESC' },
+]
+
+/**
+ * Reverse ASC to DESC, and DESC to ASC in an order clause.
+ */
+function reverseOrder<T extends { order: string }>(entries: Array<T>): Array<T> {
+  const reverse = {
+    ASC: 'DESC',
+    DESC: 'ASC',
+  }
+  return entries.map((entry) => ({ ...entry, order: reverse[entry.order] }))
+}
+
+export class ChronologicalOrder {
+  constructor(private readonly dataSource: DataSource, private readonly knexConnection: Knex) {}
+
+  async page(query: BaseQuery & Pagination): Promise<Page<StreamID>> {
+    const pagination = parsePagination(query)
+    const paginationKind = pagination.kind
+    switch (paginationKind) {
+      case PaginationKind.FORWARD: {
+        const limit = pagination.first
+        const response = await this.query<Array<Selected>>(
+          this.chronologicalForwardQuery(query, pagination)
+        )
+        const entries = response.slice(0, limit)
+        const firstEntry = entries[0]
+        const lastEntry = entries[entries.length - 1]
+        return {
+          entries: entries.map((row) => StreamID.fromString(row.stream_id)),
+          pageInfo: {
+            hasNextPage: response.length > limit,
+            hasPreviousPage: false,
+            endCursor: Cursor.stringify(asChronologicalCursor(lastEntry)),
+            startCursor: Cursor.stringify(asChronologicalCursor(firstEntry)),
+          },
+        }
+      }
+      case PaginationKind.BACKWARD: {
+        const limit = pagination.last
+        const response = await this.query<Array<Selected>>(
+          this.chronologicalBackwardQuery(query, pagination)
+        )
+        const entries = response.slice(-limit)
+        const firstEntry = entries[0]
+        const lastEntry = entries[entries.length - 1]
+        return {
+          entries: entries.map((row) => StreamID.fromString(row.stream_id)),
+          pageInfo: {
+            hasNextPage: false,
+            hasPreviousPage: response.length > limit,
+            endCursor: Cursor.stringify(asChronologicalCursor(lastEntry)),
+            startCursor: Cursor.stringify(asChronologicalCursor(firstEntry)),
+          },
+        }
+      }
+      default:
+        throw new UnsupportedOrderingError(paginationKind)
+    }
+  }
+
+  private chronologicalForwardQuery(
+    query: BaseQuery,
+    pagination: ForwardPaginationQuery
+  ): Knex.QueryBuilder {
+    const tableName = asTableName(query.model)
+    const base = this.knexConnection
+      .from(tableName)
+      .select('stream_id', 'last_anchored_at', 'created_at')
+      .orderBy(CHRONOLOGICAL_ORDER)
+      .limit(pagination.first + 1)
```

Comment:
comment for why the +1 is there please

Reply:
`// To know if we have more entries to query`.